ENT-14136 - Cross-provider key rotation

filipesoliveira opened the pull request:
Introducing the cross-provider key rotation feature, which enables nodes to safely switch between key providers (e.g., different HSM vendors or file-based keystores) without affecting access to existing states or ongoing transactions.

Although this feature is supported only in Enterprise, it relies on some Open Source logic. Additionally, due to restrictions on how classes can be imported into a CorDapp contract, part of the key rotation feature’s code is implemented here.

Some API changes have been made to `net.corda.core.contracts.Command`, `net.corda.core.contracts.CommandData`, `net.corda.core.contracts.CommandWithParties`, and `net.corda.core.crypto.SignatureMetadata`. These changes were implemented with cryptographic compatibility in mind.

The method `getMissingSigners` was updated so that any key present in the proof chain is considered during the check.

The Crypto class was updated to ensure that a transaction is considered valid only if both the transaction’s signature is valid and the proofChain in the metadata is also valid. 

Custom serialisers were added for `SignableData` and `SignatureMetadata` to preserve cryptographic compatibility with earlier versions.

The `CompatibilityTest` was updated because, although the serialized bytes of the signed transaction are not exactly the same, they do not appear to affect how the signature is generated.

The group component `KEY_ROTATION_PROOF_GROUP` was added to the enum class `ComponentGroupEnum`. This is required because a new field was added to the class `Command`. This new component is optional to ensure cryptographic compatibility with older versions of Corda.

Added a test to ensure that the serialization of `SignatureMetadata` remains unchanged after adding the new proofChain field when it is null.

Added a test to verify that the serialization of `SignableData` remains unchanged and that the generated signature is identical when the `proofChain` in the signature metadata is `null`.

The platform version has been updated. The key rotation tool will not work if the minimum platform version is not set to 170. The node will fail to start if a key rotation file is detected and the minimum platform version is not set to 170. The transaction builder will throw an error if a command contains a key proof and the minimum platform version is not set to 170.

filipesoliveira committed:
ENT-14136 - Cross-provider key rotation

filipesoliveira committed:
ENT-14136 - Cross-provider key rotation

filipesoliveira commented:
DO NOT MERGE UNTIL THE TARGET IS set to `release/os/4.15`.

filipesoliveira committed:
Updated a comment.

filipesoliveira committed:
Updated a comment.

filipesoliveira committed:
Updated a comment.

filipesoliveira committed:
Updated a comment.

filipesoliveira committed:
Fixed bug. Test was failing when the key pair was rotated more than once.

filipesoliveira committed:
Fixed bug. Test was failing when the key pair was rotated more than once.

filipesoliveira committed:
Update.

filipesoliveira committed:
Update.

filipesoliveira commented on core/src/main/kotlin/net/corda/core/contracts/CrossProviderKeyRotationProof.kt:
```diff
@@ -0,0 +1,14 @@
+        val publicKeyNew: PublicKey,
+        val signature: ByteArray,
+        val timestamp: Instant,
+        val issuer: String // Change this name to signerName
```
Remove this field.

filipesoliveira commented on core/src/main/kotlin/net/corda/core/contracts/CrossProviderKeyRotationProof.kt:
```diff
@@ -0,0 +1,14 @@
+        val publicKeyOld: PublicKey,
+        val publicKeyNew: PublicKey,
+        val signature: ByteArray,
+        val timestamp: Instant,
```
This can go too.

vlajos commented on core/src/main/kotlin/net/corda/core/transactions/TransactionWithSignatures.kt:
```diff
@@ -103,6 +106,48 @@ interface TransactionWithSignatures : NamedByHash {
+        if(requiredSigningKey.isFulfilledBy(signingKeys)) {
+            return true
+        }
+        return false
```
I think this would make it working for composite keys. (assuming newPublicKeyToOldMap has the verified relationship of keys extracted from the signatures.)

```suggestion
        val allSigningKeys = sigKeys + sigkeys.mapNotNull { newPublicKeyToOldMap[it]}
        return requiredSigningKeys.asSequence().filter { !it.isFulfilledBy(allSigningKeys) }.toSet()
    }
```

filipesoliveira replied:
Thanks, @vlajos. Your approach is much simpler and works just as well. I have updated the code based on your suggestion to support composite keys.

filipesoliveira committed:
Update.

filipesoliveira committed:
Update.

filipesoliveira committed:
Support composite keys.

filipesoliveira committed:
Support composite keys.

vlajos commented on core/src/main/kotlin/net/corda/core/crypto/internal/keyrotation/crossprovider/KeyRotationProofExtractor.kt:
```diff
@@ -0,0 +1,41 @@
+                proof.signature,
+                proof.publicKeyNew.encoded
+        )
+    }
```
Is this class working for longer proof chains?
I think if the chains have gaps, it still treats all supplied proofs as valid.

filipesoliveira replied:
Good catch. For a proof chain to be valid, it must end with the signing key, and all proofs in the metadata must be properly linked. I have updated the code.

filipesoliveira committed:
A security flaw in the validation algorithm has been addressed. All proof chains must include the key that generated the signature and must not contain any unrelated proofs.

filipesoliveira committed:
A security flaw in the validation algorithm has been addressed. All proof chains must include the key that generated the signature and must not contain any unrelated proofs.

filipesoliveira committed:
Update.

filipesoliveira committed:
Update.

filipesoliveira committed:
Update.

filipesoliveira committed:
Update.

filipesoliveira committed:
Update.

filipesoliveira committed:
Update.

filipesoliveira committed:
Abort if the proof chain contains a cycle.

filipesoliveira committed:
Abort if the proof chain contains a cycle.

filipesoliveira committed:
Fixed a bug.

filipesoliveira committed:
Fixed a bug.

filipesoliveira committed:
Updated a comment.

filipesoliveira committed:
Updated a comment.

filipesoliveira committed:
Fixed issue with amqp. The new field `_crossProviderKeyRotationProof` must be nullable to be backwards compatible.

filipesoliveira committed:
Fixed issue with amqp. The new field `_crossProviderKeyRotationProof` must be nullable to be backwards compatible.

filipesoliveira committed:
Update

filipesoliveira committed:
Update

filipesoliveira committed:
Update

filipesoliveira committed:
Update

filipesoliveira committed:
Added the key proof map to the `CommandWithParties` and `Command`.

filipesoliveira committed:
Added the key proof map to the `CommandWithParties` and `Command`.

filipesoliveira committed:
Updated the method `resolveToCurrentParty` so only one call is made to the identity service.

filipesoliveira committed:
Updated the method `resolveToCurrentParty` so only one call is made to the identity service.

filipesoliveira committed:
Update resolveToCurrentParty to use `wellKnownPartyFromAnonymous`.

filipesoliveira committed:
Update resolveToCurrentParty to use `wellKnownPartyFromAnonymous`.

filipesoliveira committed:
Renamed `keyRotationProofs` to keyRotationProofChainMap

filipesoliveira committed:
Renamed `keyRotationProofs` to keyRotationProofChainMap

filipesoliveira committed:
Update

filipesoliveira committed:
Update

filipesoliveira committed:
Revert changes made.

filipesoliveira committed:
Revert changes made.

filipesoliveira committed:
Revert changes made.

filipesoliveira committed:
Revert changes made.

filipesoliveira committed:
Revert changes made.

filipesoliveira committed:
Revert changes made.

filipesoliveira committed:
Added exceptions to prevent the use of certain key rotation features in Corda OS.

filipesoliveira committed:
Added exceptions to prevent the use of certain key rotation features in Corda OS.

filipesoliveira committed:
Reverted changes

filipesoliveira committed:
Reverted changes

filipesoliveira committed:
Updated a comment

filipesoliveira committed:
Updated a comment

filipesoliveira committed:
Updated.

filipesoliveira committed:
Updated.

filipesoliveira committed:
Reverted change.

filipesoliveira committed:
Reverted change.

filipesoliveira committed:
Update

filipesoliveira committed:
Update

filipesoliveira committed:
Reverted a change

filipesoliveira committed:
Reverted a change

filipesoliveira committed:
Added a check to prevent people from using the key rotation features in Corda OS to avoid unexpected behaviour.

filipesoliveira committed:
Added a check to prevent people from using the key rotation features in Corda OS to avoid unexpected behaviour.

filipesoliveira committed:
Update.

filipesoliveira committed:
Update.

filipesoliveira committed:
Update.

filipesoliveira committed:
Update.

filipesoliveira committed:
Added tests and a comment.

filipesoliveira committed:
Added tests and a comment.

filipesoliveira committed:
Added tests.

filipesoliveira committed:
Added tests.

filipesoliveira committed:
Added tests.

filipesoliveira committed:
Added tests.

filipesoliveira committed:
Added tests.

filipesoliveira committed:
Added tests.

filipesoliveira committed:
Updated SignatureMetadata so it is backwards compatible.

filipesoliveira committed:
Updated SignatureMetadata so it is backwards compatible.

filipesoliveira committed:
Fixed issues reported by detekt.

filipesoliveira committed:
Fixed issues reported by detekt.

filipesoliveira committed:
Update the file api-current.txt.

filipesoliveira committed:
Update the file api-current.txt.

filipesoliveira committed:
Fixed a bug that caused the Merkle tree to differ after deserialization.

filipesoliveira committed:
Fixed a bug that caused the Merkle tree to differ after deserialization.

filipesoliveira committed:
Fixed bug.

filipesoliveira committed:
Fixed bug.

filipesoliveira committed:
Fixed issue reported by detekt.

filipesoliveira committed:
Fixed issue reported by detekt.

filipesoliveira committed:
Update api.

filipesoliveira committed:
Update api.

filipesoliveira committed:
Fixed a test. `NetworkParamenters` are no longer the last component in the list of component groups.

filipesoliveira committed:
Fixed a test. `NetworkParamenters` are no longer the last component in the list of component groups.

filipesoliveira committed:
Fixed bug.

filipesoliveira committed:
Fixed bug.

filipesoliveira committed:
Fixed bug.

filipesoliveira committed:
Fixed bug.

filipesoliveira committed:
Fixed tests and a bug.

filipesoliveira committed:
Fixed tests and a bug.

filipesoliveira committed:
Fixed tests.

filipesoliveira committed:
Fixed tests.

filipesoliveira committed:
Fixed issue with serialization.

filipesoliveira committed:
Fixed issue with serialization.

filipesoliveira committed:
Updated the api.

filipesoliveira committed:
Updated the api.

filipesoliveira committed:
Trying a different approach to keep compatibility with older versions of Corda 4.x.

filipesoliveira committed:
Trying a different approach to keep compatibility with older versions of Corda 4.x.

filipesoliveira committed:
Trying a different approach to keep compatibility with older versions of Corda 4.x.

filipesoliveira committed:
Trying a different approach to keep compatibility with older versions of Corda 4.x.

filipesoliveira committed:
Added a custom serializer for SignatureMetadata to ensure that the generated bytes match the previous format exactly, byte for byte. This will allow older versions of Corda 4.x to interoperate with newer versions.

filipesoliveira committed:
Added a custom serializer for SignatureMetadata to ensure that the generated bytes match the previous format exactly, byte for byte. This will allow older versions of Corda 4.x to interoperate with newer versions.

filipesoliveira committed:
Update to the serializer.

filipesoliveira committed:
Update to the serializer.

filipesoliveira committed:
Added a customer serializer for SignableDataSerializer class.

filipesoliveira committed:
Added a customer serializer for SignableDataSerializer class.

filipesoliveira committed:
Fixed a JSON serialization issue.

filipesoliveira committed:
Fixed a JSON serialization issue.

filipesoliveira committed:
Updated the class KeyRotationProof to ensure the equals method works as expected.

filipesoliveira committed:
Updated the class KeyRotationProof to ensure the equals method works as expected.

filipesoliveira committed:
Updated the serialization again. The serializer was missing.

filipesoliveira committed:
Updated the serialization again. The serializer was missing.

filipesoliveira committed:
Fixed issue reported by detekt.

filipesoliveira committed:
Fixed issue reported by detekt.

filipesoliveira committed:
Updated the test.

filipesoliveira committed:
Updated the test.

filipesoliveira committed:
Fixed build warning

filipesoliveira committed:
Fixed build warning

filipesoliveira committed:
Small change.

filipesoliveira committed:
Small change.

filipesoliveira committed:
Updated a comment.

filipesoliveira committed:
Updated a comment.

filipesoliveira committed:
Add the KEY_ROTATION_PROOF_GROUP iff one of the commands sets `keyRotationProofChainMap` to a non-null value.

filipesoliveira committed:
Add the KEY_ROTATION_PROOF_GROUP iff one of the commands sets `keyRotationProofChainMap` to a non-null value.

filipesoliveira committed:
Reverted changes made to the test `PartialMerkleTreeTest` and `PartialMerkleTreeWithNamedHashTest`.

filipesoliveira committed:
Reverted changes made to the test `PartialMerkleTreeTest` and `PartialMerkleTreeWithNamedHashTest`.

filipesoliveira committed:
Reverted changes made to the test `PartialMerkleTreeTest`, `PartialMerkleTreeWithNamedHashTest`, and `PartialMerkleTreeWithNamedHashMultiAlgTreeTest`.

filipesoliveira committed:
Reverted changes made to the test `PartialMerkleTreeTest`, `PartialMerkleTreeWithNamedHashTest`, and `PartialMerkleTreeWithNamedHashMultiAlgTreeTest`.

filipesoliveira committed:
The function `dummyCommand` now creates a command with the keyRotationProof set to null instead of an empty map.

filipesoliveira committed:
The function `dummyCommand` now creates a command with the keyRotationProof set to null instead of an empty map.

filipesoliveira committed:
Added an assert to the test `issueCashTransactionV3ReadTest`.

filipesoliveira committed:
Added an assert to the test `issueCashTransactionV3ReadTest`.

filipesoliveira committed:
Reverted the changes made to the test `CompatibleTransactionTests`.

filipesoliveira committed:
Reverted the changes made to the test `CompatibleTransactionTests`.

filipesoliveira committed:
Small changes to the ContractDSL.kt file.

filipesoliveira committed:
Small changes to the ContractDSL.kt file.

filipesoliveira committed:
Removed extra space.

filipesoliveira committed:
Removed extra space.

filipesoliveira committed:
Added the test `CompatibleTransactionWithKeyRotationChainProofTests`, which replicates the behavior of `CompatibleTransactionTest` but uses a key rotation proof map.
Updated `CompatibleTransactionTests` to verify that the key rotation proof group is absent when the key rotation proof in the command is set to null.

filipesoliveira committed:
Added the test `CompatibleTransactionWithKeyRotationChainProofTests`, which replicates the behavior of `CompatibleTransactionTest` but uses a key rotation proof map.
Updated `CompatibleTransactionTests` to verify that the key rotation proof group is absent when the key rotation proof in the command is set to null.

filipesoliveira committed:
Updated the method Command.toString so it includes the key rotation proofs  when it is not null.

filipesoliveira committed:
Updated the method Command.toString so it includes the key rotation proofs  when it is not null.

filipesoliveira committed:
Added the test TransactionBuilderWithKeyRotationProofTest, which replicates the behavior of TransactionBuilderTest but uses a key rotation proof map.

filipesoliveira committed:
Added the test TransactionBuilderWithKeyRotationProofTest, which replicates the behavior of TransactionBuilderTest but uses a key rotation proof map.

filipesoliveira committed:
Added the test "building Merkle tree for a tx with key rotation proof and for nonce test", which replicates the behavior of "building Merkle tree for a tx and nonce test" but uses a key rotation proof map.

filipesoliveira committed:
Added the test "building Merkle tree for a tx with key rotation proof and for nonce test", which replicates the behavior of "building Merkle tree for a tx and nonce test" but uses a key rotation proof map.

filipesoliveira committed:
Fixed issue reported by detekt.

filipesoliveira committed:
Fixed issue reported by detekt.

filipesoliveira committed:
Added a test to ensure that the serialization of SignatureMetadata remains unchanged after adding the new proofChain field when it is null.
Added a test to verify that the serialization of SignableData remains unchanged and that the generated signature is identical when the proofChain in the signature metadata is null.

The files added to the resources were generated by Corda 4.14 without applying any of the changes.

filipesoliveira committed:
Added a test to ensure that the serialization of SignatureMetadata remains unchanged after adding the new proofChain field when it is null.
Added a test to verify that the serialization of SignableData remains unchanged and that the generated signature is identical when the proofChain in the signature metadata is null.

The files added to the resources were generated by Corda 4.14 without applying any of the changes.

filipesoliveira committed:
Revert change to the api (public interface net.corda.core.contracts.NotaryInstruction has been removed.)

filipesoliveira committed:
Revert change to the api (public interface net.corda.core.contracts.NotaryInstruction has been removed.)

filipesoliveira committed:
Updated the `initiateFlow` method to ensure that the Party being used always contains the latest details.

filipesoliveira committed:
Updated the `initiateFlow` method to ensure that the Party being used always contains the latest details.

filipesoliveira committed:
Revert "Updated the `initiateFlow` method to ensure that the Party being used always contains the latest details."

This reverts commit c43c57e017685e409390ac16077b731bdea812db.

filipesoliveira committed:
Revert "Updated the `initiateFlow` method to ensure that the Party being used always contains the latest details."

This reverts commit c43c57e017685e409390ac16077b731bdea812db.

filipesoliveira committed:
Reapply "Updated the `initiateFlow` method to ensure that the Party being used always contains the latest details."

This reverts commit 14328b2b8a03e8077f21eaea09f76dd2cde8ff5d.

filipesoliveira committed:
Reapply "Updated the `initiateFlow` method to ensure that the Party being used always contains the latest details."

This reverts commit 14328b2b8a03e8077f21eaea09f76dd2cde8ff5d.

filipesoliveira committed:
Fixed a test. It was falling because when the payment subflow raised an exception at while calling `initiateFlow`, the balance of Alice would be reverted to zero again.

filipesoliveira committed:
Fixed a test. It was falling because when the payment subflow raised an exception at while calling `initiateFlow`, the balance of Alice would be reverted to zero again.

filipesoliveira committed:
Updated the platform version.

# Conflicts:
#	constants.properties

filipesoliveira committed:
Updated the platform version.

filipesoliveira committed:
Updated a test that was failing.

filipesoliveira committed:
Updated a test that was failing.

filipesoliveira committed:
Updated another test that was failing.

filipesoliveira committed:
Updated another test that was failing.

filipesoliveira committed:
Removed unused import.

filipesoliveira committed:
Removed unused import.

filipesoliveira committed:
Fixed a bug that was causing issues when anonymous keys were being used.

filipesoliveira committed:
Fixed a bug that was causing issues when anonymous keys were being used.

adelel1 commented on core/src/main/kotlin/net/corda/core/crypto/keyrotation/crossprovider/KeyRotationProofChain.kt:
```diff
@@ -0,0 +1,116 @@
+            Crypto.doVerify(scheme, proof.publicKeyOld, proof.signature, proof.publicKeyNew.encoded)
+            true
+        } catch (_: GeneralSecurityException) {
+            false
```
Shouldn't we log the exception here?

filipesoliveira replied:
A log message has been added in.

adelel1 commented on core/src/main/kotlin/net/corda/core/crypto/keyrotation/crossprovider/PartyIdentityResolver.kt:
```diff
@@ -0,0 +1,98 @@
+     *
+     * If no proof chain is available or validation fails, returns false.
+     */
+    private fun isSameKey(originalKey: PublicKey, candidateKey: PublicKey): Boolean {
```
Doesnt this only work if candidateKey is the end of the proof chain.
e.g.
say you have a chain A -> B -> C -> D -> E
A and E can be considered the same. But what about A and C, wouldn't this func return false if C is passed as candidate key. Dont we want this returning true for A and C as well?

filipesoliveira replied:
_Doesnt this only work if candidateKey is the end of the proof chain._
Yes, it will only work if the candidate key is at the end of the proof chain. 

_Dont we want this returning true for A and C as well?_
No, not in that scenario. If key C is being verified, there will be a corresponding proof chain that starts with A and ends with C. The current implementation requires the proof chain to start with the original key and end with the candidate key. This simplifies the logic used to locate a proof chain and avoids the additional complexity of merging multiple proof chains that share common proofs.

adelel1 commented on core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt:
```diff
@@ -163,8 +163,14 @@ abstract class FlowLogic<out T> {
+    fun initiateFlow(party: Party): FlowSession {
+        // The method `wellKnownPartyFromAnonymous` is called in case the party passed in contains a key that has been rotated. In such a case,
+        // the well known party will be returned with the new key, and the flow session will be initiated with that new key.
+        val upToDateParty = serviceHub.identityService.wellKnownPartyFromAnonymous(party)
```
Just wondering on the performance impact of this, e.g if no rotations have taken place on the node we don't need to do this.

adelel1 replied:
Is there an in memory cache we can use to determine if need to attempt this at all.

filipesoliveira replied:
_Is there an in memory cache we can use to determine if need to attempt this at all._
Unfortunately, the short answer is no. Corda OS does not store key rotation proofs in the database. However, the method `wellKnownPartyFromAnonymous` uses caches from the `NamedCacheFactory` to perform lookups. And from what I recall from the meetings, it was agreed that this would be sufficient. 

If performance becomes a concern, I can revert this change and document that CorDapps must pass the latest party when initiating the flow, as was required previously. This improvement is more of a nice-to-have for the CorDapps developers.

adelel1 commented on core/src/main/kotlin/net/corda/core/internal/TransactionUtils.kt:
```diff
@@ -170,7 +173,18 @@ fun deserialiseCommands(
+    val keyRotationProofList: List<Map<PublicKey, KeyRotationProofChain>> =
+            if (keyRotationProofRaw.isEmpty()) {
+                // No group present (older wire transaction); default to empty map per signer entry to preserve indexing.
+                List(signersList.size) { emptyMap<PublicKey, KeyRotationProofChain>() }
```
Do we need to create this, later you check for isEmpty and then return null.

filipesoliveira replied:
This code has been simplified. The block `if (keyRotationProofRaw.isEmpty())` is no longer required.

adelel1 commented on core/src/main/kotlin/net/corda/core/internal/TransactionUtils.kt:
```diff
@@ -226,6 +242,14 @@ fun createComponentGroups(inputs: List<StateRef>,
+
+    if(commands.any { it.keyRotationProofChainMap != null }) {
+        // Adding key rotation proof map to its own group. If all the commands have null key rotation proof map,
+        // we skip adding the group to ensure compatibility with older versions of Corda which don't have this group.
```
Makes sense to not include the group if we dont need to. 
But from a backwards compatibility point of view it can be included, it wont affect backwards compatibility.  The older node just wont know about the group, so wont be attempting to deserialize or use.

filipesoliveira replied:
Yes, adding a new group will not affect deserialization. However, it will change the Merkle tree of the transaction, and nodes running an older version of Corda will not be able to verify the transaction signature.

Below is a screenshot of the code that calculates the Merkle tree:
<img width="2206" height="1522" alt="image" src="https://github.com/user-attachments/assets/8fc1cb9e-f22e-4a62-90de-6c8935bf564d" />

adelel1 commented on core/src/main/kotlin/net/corda/core/transactions/TransactionWithSignatures.kt:
```diff
@@ -100,7 +101,8 @@ interface TransactionWithSignatures : NamedByHash {
      */
     fun getMissingSigners(): Set<PublicKey> {
-        val sigKeys = sigs.mapToSet { it.by }
+        val sigKeys = sigs.mapToSet { it.by } + sigs.getKeyLineage()
```
Is this right? This is saying every public key in the proof chain for the sigs is treated as missing.

filipesoliveira replied:
Yes, the code change is correct. The `sigKeys` set represents all keys that have signed the transaction. The `sigs` list contains the actual keys that have signed over the transaction. The method `sigs.getKeyLineage()` returns all public keys in the proof chain, which can be interpreted as having implicitly signed the transaction via a key present in sigs.

The code `requiredSigningKeys.asSequence().filter { !it.isFulfilledBy(sigKeys) }.toSet()` checks whether any required signing key has not signed the transaction. In other words, it verifies if any key in `requiredSigningKeys` is missing from `sigKeys`, where `sigKeys` represents the set of keys that have signed the transaction.

adelel1 replied:
Ok, got it. I guess i missed the second line!

adelel1 commented on node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt:
```diff
@@ -154,4 +158,16 @@ class InMemoryIdentityService(
+
+    override fun containsProofChain(publicKey: PublicKey): Boolean {
+        throw NotSupportedException("Cross-Provider key rotation is not supported by Corda OS")
+    }
```
Can you provide these defaults on the interface. Maybe update the message slightly. i.e. so you only change whats in core.

filipesoliveira replied:
Done. The exception was moved to the interface.

adelel1 commented on node/src/main/kotlin/net/corda/node/services/identity/PersistentIdentityService.kt:
```diff
@@ -512,6 +515,18 @@ class PersistentIdentityService(cacheFactory: NamedCacheFactory) : SingletonSeri
+    override fun containsProofChain(publicKey: PublicKey): Boolean {
+        throw NotSupportedException("Cross-Provider key rotation is not supported by Corda OS")
+    }
+
```
Can you provide these defaults on the interface. Maybe update the message slightly. i.e. so you only change whats in core.

adelel1 commented on node/src/main/kotlin/net/corda/node/services/keys/BasicHSMKeyManagementService.kt:
```diff
@@ -110,6 +110,10 @@ class BasicHSMKeyManagementService(
+    override fun isKeyRotated(candidateKey: PublicKey): Boolean {
+        return false
+    }
+
```
Can you provide these defaults on the interface.  i.e. so you only change whats in core.

filipesoliveira replied:
Done.

adelel1 commented on testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/MockKeyManagementService.kt:
```diff
@@ -39,6 +39,8 @@ class MockKeyManagementService(
     override fun filterMyKeys(candidateKeys: Iterable<PublicKey>): Iterable<PublicKey> = candidateKeys.filter { it in this.keys }
 
+    override fun isKeyRotated(candidateKey: PublicKey): Boolean { return false }
+
```
Dont need this if default is in interface.

filipesoliveira replied:
Done.

filipesoliveira committed:
A message is now logged when a proof is invalid.

filipesoliveira committed:
Refactored the code to remove the auxiliary variable.

filipesoliveira committed:
Move the `NotSupportedException("Cross-Provider key rotation is not supported")` to the interface level.

filipesoliveira committed:
Fixed compilation issue.

filipesoliveira committed:
Moved the default implementation of the method `isKeyRotated` to the interface.

filipesoliveira committed:
Updated the file 'api-current.txt'.

filipesoliveira committed:
Reverted the changes in JacksonSupport.kt, CordaModule.kt, and JacksonSupportTest.kt as requested in the PR. The remaining changes are required.

filipesoliveira commented on finance/workflows/src/test/kotlin/net/corda/finance/flows/CompatibilityTest.kt:
```diff
@@ -31,8 +32,8 @@ class CompatibilityTest {
-	fun issueCashTansactionReadTest() {
-        val inputStream = javaClass.classLoader.getResourceAsStream("compatibilityData/v3/node_transaction.dat")
+	fun issueCashTransactionReadTest() {
+        val inputStream = javaClass.classLoader.getResourceAsStream("compatibilityData/v4/node_transaction.dat")
```
This change is required because the serialisation of the transaction is not an exact match.

filipesoliveira commented on finance/workflows/src/test/kotlin/net/corda/finance/flows/CompatibilityTest.kt:
```diff
@@ -57,6 +58,33 @@ class CompatibilityTest {
     }
 
+    @Test(timeout=300_000)
+	fun issueCashTransactionV3ReadTest() {
```
This change ensures that, although the transaction serialization is not identical, it remains sufficiently consistent for cryptographic signature verification. This is only possible because the portions of the serialization that differ are not included in the data used during transaction signing.

filipesoliveira replied:
This is important to ensure compatibility, allowing nodes running older versions of Corda to continue processing transactions produced by nodes running the latest changes.

filipesoliveira commented on node/src/integration-test/kotlin/net/corda/node/services/identity/TrustRootTest.kt:
```diff
@@ -108,7 +110,9 @@ class TrustRootTest {
         assertEquals(rootCa2.certificate, bob.services.identityService.trustRoot)
 
-        alice.services.startFlow(CashIssueAndPaymentFlow(1000.DOLLARS, ref, bob.party, false, notary.party))
+        alice.services.startFlow(CashIssueFlow(1000.DOLLARS, ref, notary.party))
```
This change is necessary because an exception now causes the entire flow to roll back. As a result, the `CashIssue` subflow invoked within `CashIssueAndPaymentFlow` no longer takes effect, causing the test to fail.

Separating the flows ensures that the cash issuance flow completes successfully while still preserving the expected failure behaviour of CashPaymentFlow.

adelel1 replied:
@filipesoliveira What was the change that caused 

"because an exception now causes the entire flow to roll back. "

Dont see how this is related to key rotation.
Wouldn't this also affect customer code?

filipesoliveira replied:
The code fails because of this change:
<img width="1252" height="219" alt="image" src="https://github.com/user-attachments/assets/014a52a0-c639-49a7-b422-412cc26ed9ef" />

That change is a nice-to-have. We may want to revert it to be on the safe side.

filipesoliveira replied:
Yes, this could impact customers if they are using the subflow.

filipesoliveira replied:
Another solution is to not throw an exception if the method returns null. This is actually also a good approach.

filipesoliveira replied:
I will revert the changes made to the tests and update the code so that it does not throw an exception when the method returns `null`.

I am still open to reverting the change made to `initiateFlow` if that is the preferred approach.

filipesoliveira replied:
Done.

filipesoliveira committed:
Reverted a change as requested in the PR.

filipesoliveira committed:
Reverted a change as requested in the PR.

filipesoliveira commented on constants.properties:
```diff
@@ -13,7 +13,7 @@ internalPublishVersion=1.+
 # net.corda.core.internal.CordaUtilsKt.PLATFORM_VERSION as well. #
 # ***************************************************************#
-platformVersion=160
+platformVersion=170
```
This change is required to allow cross-provider key rotation to determine whether all nodes in the network support the feature.

adelel1 replied:
For every major release the platform version gets bumped.

filipesoliveira committed:
Rewrote the test in partialMerkleTree to avoid modifying additional files.

filipesoliveira committed:
Simplified the implementation of `getKeyLineage`.

filipesoliveira committed:
Updated a comment for `Command` data class.

filipesoliveira committed:
Updated a comment for `filter` method.

filipesoliveira committed:
Updated the error message in IdentityService.

filipesoliveira committed:
Added a `require` to prevent `keyRotationProofChainMap` from being empty. Null should be used instead in those cases.

adelel1 commented on core/src/main/kotlin/net/corda/core/contracts/Structures.kt:
```diff
@@ -192,18 +193,26 @@ abstract class TypeOnlyCommandData : CommandData {
     private fun commandDataToString() = value.toString().let { if (it.contains("@")) it.replace('$', '.').split("@")[0] else it }
-    override fun toString() = "${commandDataToString()} with pubkeys ${signers.joinToString { it.toStringShort() }}"
+    override fun toString(): String {
+        if(keyRotationProofChainMap == null || keyRotationProofChainMap.isEmpty()) {
```
You can use 
if(keyRotationProofChainMap.isNullOrEmpty()) {
instead .

filipesoliveira replied:
`isNullorEmpty()` does not work here. The code does not compile:

`core/src/main/kotlin/net/corda/core/contracts/Structures.kt:212:37: error: unresolved reference. None of the following candidates is applicable because of receiver type mismatch: 
@InlineOnly public inline fun CharSequence?.isNullOrEmpty(): Boolean defined in kotlin.text
        if(keyRotationProofChainMap.isNullOrEmpty()) {`

filipesoliveira committed:
Updated the code to use `isNullOrEmpty` instead.

adelel1 commented on core/src/main/kotlin/net/corda/core/crypto/Crypto.kt:
```diff
@@ -570,7 +570,20 @@ object Crypto {
+        }
+        val verificationResult = signatureMetadata.proofChain.isValid(publicKey)
+        if (!verificationResult) {
+            throw SignatureException("Signature Verification failed!")
```
Can isValid either return true, or an exception describing the verification failure, so we get more info on what failed. And maybe have in the Exception some details of the failure?

filipesoliveira replied:
I have added log messages to the `isValid` method. The method should return true or false instead of throwing an exception. Otherwise, returning a boolean is not meaningful and the method should probably be renamed to something else.

filipesoliveira committed:
Revert "Updated the code to use `isNullOrEmpty` instead."

This reverts commit 52dfe890f123c1115f94cc99af4005b66aaaec3a.

adelel1 commented on core/src/main/kotlin/net/corda/core/node/ServiceHub.kt:
```diff
@@ -308,7 +308,19 @@ interface ServiceHub : ServicesForResolution {
      */
-    fun signInitialTransaction(builder: TransactionBuilder): SignedTransaction = signInitialTransaction(builder, legalIdentityKey)
+    fun signInitialTransaction(builder: TransactionBuilder): SignedTransaction {
+        val requiredSigners = builder.commands().flatMap { it.signers }.toSet()
```
We're always going to do this, even if nothing is rotated in the node. Wondering if we can first check if rotation is a possibility if so proceed, otherwise do as we originally did - i.e. sign with legalIdentityKey

filipesoliveira replied:
I will have a look at this. It should be possible since we have access to the `keyManagementService`.

filipesoliveira replied:
Done. The code has been updated so that if the node has not rotated its key, the additional checks are never performed.

adelel1 commented on core/src/main/kotlin/net/corda/core/node/ServiceHub.kt:
```diff
@@ -308,7 +308,19 @@ interface ServiceHub : ServicesForResolution {
+            // so we need to check if any of the required keys is an old key that has been rotated and if so, sign with the new key instead.
+            for (requiredSigner in requiredSigners) {
+                if (keyManagementService.isKeyRotated(requiredSigner)) {
+                    return signInitialTransaction(builder, requiredSigner)
```
The comment says sign with the "new key" but your signing wit the requiredSigner?
Also this is just going to sign with the first rotated required signer it finds. Shouldn't it also check that legalIdentityKey and requiredSigner are part of same chain.

filipesoliveira replied:
_The comment says sign with the "new key" but your signing wit the requiredSigner?_
Thanks for spotting that. That was a mistake. I meant to say “sign with that key” rather than “sign with the new key.

_Also this is just going to sign with the first rotated required signer it finds. Shouldn't it also check that legalIdentityKey and requiredSigner are part of same chain._
Each node has only one key rotation chain, and the `KeyManagementService` only contains keys that belong to that node. Therefore, if the required key has been rotated, it must belong to the same key rotation chain as the legalIdentityKey. This assumption may need to be revisited in the future to support confidential identities. However, for now, this should not be an issue.

What scenarios are you considering where `legalIdentityKey` and `requiredSigner` would not belong to the same key rotation chain?

filipesoliveira committed:
Added logs when to `isValid` method to provide more details when the rotation between two keys is not valid.

filipesoliveira committed:
Update the method `signInitialTransaction` to avoid the extra checks when the node has not rotated its key.

filipesoliveira committed:
Updated a comment.

filipesoliveira committed:
Fixed issue reported by detekt

filipesoliveira committed:
Added a check to make sure that the sizes of CommandData and keyRotationProofList match or the keyRotationProofList is empty.

filipesoliveira committed:
Fixed a typo.

filipesoliveira committed:
Added a comment to the class `PartyIdentityResolver` to tell CorDapp developers how to use the class.

filipesoliveira committed:
Renamed the test `KeyRotationProofTest` to `KeyRotationProofChainTest`

filipesoliveira committed:
Fixed a typo. "they is no" -> "there is no"

filipesoliveira committed:
Do not throw an exception in `initiateFlow` when `wellKnownPartyFromAnonymous` returns null.
Revert the change made to the test `TrustRootTest`.

filipesoliveira committed:
Updated the method `generateProofChainMap` to return null if the map is empty.

filipesoliveira committed:
Updated the API.

filipesoliveira committed:
Removed check since keyRotationProofChainMap cannot be empty.

filipesoliveira committed:
Replaced Map for a SortedMap in the command.

filipesoliveira committed:
Added a test that verifies Corda OS can validate a transaction with a key rotation proof.

filipesoliveira committed:
Updated a comment

filipesoliveira committed:
Removed extra spaces and fixed mis-indentation.

filipesoliveira committed:
Replaced the check `if(map.keys.firstOrNull() is PublicKey && map.values.firstOrNull() is KeyRotationProofChain)` to `if(map.keys.all { it is PublicKey } && map.values.all { it is KeyRotationProofChain } )`

adelel1 commented on core/src/main/kotlin/net/corda/core/crypto/keyrotation/crossprovider/KeyRotationProofChain.kt:
```diff
@@ -0,0 +1,142 @@
+        }
+        
+        if (isEmpty()) {
+            logger.info("Validation failed. Chain is empty")
```
Wondering if all the log.info in this func should be log.warn instead? ie. we don't expect them to occur do we?
Also for the log message do you want to output the key as well. So that we have something to debug with if this message is seen, presumably original key? Same for other log lines in this func.

filipesoliveira replied:
The INFO log messages in `KeyRotationProofChain` have been changed to WARN level. The logs now also include both the original key and the current key to make it easier to diagnose and troubleshoot potential issues. Commit: 64dd321c

filipesoliveira committed:
The INFO log messages in `KeyRotationProofChain` have been changed to WARN level. The logs now also include both the original key and the current key to make it easier to diagnose and troubleshoot potential issues.
